### PR TITLE
[FEATURE] complete xQuest XML format integration, XFDR

### PR DESCRIPTION
**Pull request by enetz**
xquest.xml files are now well integrated with the internal structures of OpenMS,
including conversion into mzIdentML and idXML with IDFileConverter.

XFDR, the Cross-Linking MS FDR calculation tool, was adapted to treat all possible ID input file types equally (idXML, mzIdentML and xquest.xml).
Additional parameters for a user defined decoy string and xquest.xml output were added to XFDR as well.

### CODE REVIEW

File: src/openms/include/OpenMS/FORMAT/HANDLERS/XQuestResultXMLHandler.h [hunk 1]
```diff
@@ -92,23 +97,38 @@ namespace OpenMS
        */
       UInt getNumberOfHits() const;
 
+      //Docu in base class
+      virtual void writeTo(std::ostream& os) override;
+
     private:
 
+      /// Progress logger
+      //const ProgressLogger& logger_;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/HANDLERS/XQuestResultXMLHandler.h]: remove and add as TODO?

File: src/openms/include/OpenMS/FORMAT/HANDLERS/XQuestResultXMLHandler.h [hunk 2]
```diff
@@ -92,23 +97,38 @@ namespace OpenMS
...
 
       UInt n_hits_; // Total no. of hits found in the result XML file
 
       // Keeps track of the minscore and maxscore encountered
       double min_score_;
       double max_score_;
 
-      Size min_n_ions_per_spectrum_;
-      bool load_to_peptideHit_;  // Whether Meta data of peptide identification should also be loaded to peptide hit
+      // Size min_n_hits_per_spectrum_;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/HANDLERS/XQuestResultXMLHandler.h]: remove or add as TODO (and line below)

File: src/openms/include/OpenMS/FORMAT/HANDLERS/XQuestResultXMLHandler.h [hunk 3]
```diff
@@ -121,11 +141,20 @@ namespace OpenMS
...
       double rt_light_;
+      double rt_heavy_;
+
+      // Current experimental m/z of spectrum pair
+      double mz_light_;
+      double mz_heavy_;
 
       // The masses of the Monolinks
-      std::set< double > monolinks_masses_;
+      // DoubleList monolinks_masses_;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/HANDLERS/XQuestResultXMLHandler.h]: remove?

File: src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp [hunk 4]
```diff
@@ -701,33 +702,42 @@ namespace OpenMS
         {
           seq_alpha.setModification(alpha_pos, mods[0]);
           mod_set = true;
+          ph_alpha.setMetaValue("xl_mod", mods[0]);
+          // ph_alpha.setMetaValue("xl_mass", DataValue(top_csms_spectrum[i].cross_link.cross_linker_mass));
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp]: if this is valuable debug code consider adding something like: #ifdef DEBUG_OPXL ...
to enable it

File: src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp [hunk 5]
```diff
@@ -701,33 +702,42 @@ namespace OpenMS
...
         }
         if (!mod_set) // If no equivalent mono-link exists in the UNIMOD or XLMOD databases, use the given name to construct a placeholder
         {
           String mod_name = String("unknown mono-link " + top_csms_spectrum[i].cross_link.cross_linker_name + " mass " + String(top_csms_spectrum[i].cross_link.cross_linker_mass));
           //seq_alpha.setModification(alpha_pos, mod_name);
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp]: remove?

File: src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp [hunk 6]
```diff
@@ -701,33 +702,42 @@ namespace OpenMS
...
-          LOG_DEBUG << "unknown mono-link" << endl;
+          // LOG_DEBUG << "unknown mono-link" << endl;
           ph_alpha.setMetaValue("xl_mod", mod_name);
-          ph_alpha.setMetaValue("xl_mass", DataValue(top_csms_spectrum[i].cross_link.cross_linker_mass));
+          // ph_alpha.setMetaValue("xl_mass", DataValue(top_csms_spectrum[i].cross_link.cross_linker_mass));
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp]: see above (also applies to commented out section below)

File: src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp [hunk 7]
```diff
@@ -701,33 +702,42 @@ namespace OpenMS
...
 
       if (top_csms_spectrum[i].cross_link.getType() == OPXLDataStructs::LOOP)
       {
         ph_alpha.setMetaValue("xl_pos2", DataValue(beta_pos));
       }
 
       // Error calculation
-      double weight = seq_alpha.getMonoWeight() + top_csms_spectrum[i].cross_link.cross_linker_mass;
+      double weight = seq_alpha.getMonoWeight();
       if (top_csms_spectrum[i].cross_link.getType() == OPXLDataStructs::CROSS)
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp]: maybe add some comments how the error is calculated

File: src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp [hunk 8]
```diff
@@ -780,23 +792,31 @@ namespace OpenMS
...
-      ph_alpha.setMetaValue("OpenPepXL:HyperAlpha", top_csms_spectrum[i].HyperAlpha);
-      ph_alpha.setMetaValue("OpenPepXL:HyperBeta", top_csms_spectrum[i].HyperBeta);
-      ph_alpha.setMetaValue("OpenPepXL:HyperBoth",top_csms_spectrum[i].HyperBoth);
+      ph_alpha.setMetaValue("HyperCommon",top_csms_spectrum[i].HyperCommon);
+      ph_alpha.setMetaValue("HyperXlink",top_csms_spectrum[i].HyperXlink);
+      ph_alpha.setMetaValue("HyperAlpha", top_csms_spectrum[i].HyperAlpha);
+      ph_alpha.setMetaValue("HyperBeta", top_csms_spectrum[i].HyperBeta);
+      ph_alpha.setMetaValue("HyperBoth",top_csms_spectrum[i].HyperBoth);
+
+      // ph_alpha.setMetaValue("OpenPepXL:PScoreCommon",top_csms_spectrum[i].PScoreCommon);
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp]: remove or consider adding as TODO / debug option

File: src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp [hunk 9]
```diff
@@ -828,23 +850,25 @@ namespace OpenMS
...
-        ph_beta.setMetaValue("OpenPepXL:PScoreAlpha",top_csms_spectrum[i].PScoreAlpha);
-        ph_beta.setMetaValue("OpenPepXL:PScoreBeta",top_csms_spectrum[i].PScoreBeta);
-        ph_beta.setMetaValue("OpenPepXL:PScoreBoth",top_csms_spectrum[i].PScoreBoth);
+        ph_beta.setMetaValue("HyperCommon",top_csms_spectrum[i].HyperCommon);
+        ph_beta.setMetaValue("HyperXlink",top_csms_spectrum[i].HyperXlink);
+        ph_beta.setMetaValue("HyperAlpha",top_csms_spectrum[i].HyperAlpha);
+        ph_beta.setMetaValue("HyperBeta",top_csms_spectrum[i].HyperBeta);
+        ph_beta.setMetaValue("HyperBoth",top_csms_spectrum[i].HyperBoth);
+
+        // ph_beta.setMetaValue("OpenPepXL:PScoreCommon",top_csms_spectrum[i].PScoreCommon);
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp]: same here

File: src/openms/source/FORMAT/XQuestResultXMLFile.cpp [hunk 10]
```diff
@@ -50,13 +51,13 @@ namespace OpenMS
   }
 
   void XQuestResultXMLFile::load(const String & filename,
-                                 std::vector< std::vector < PeptideIdentification > > & csms,
-                                 std::vector< ProteinIdentification > & prot_ids,
-                                 Size min_n_hits_per_spectrum,
-                                 bool load_to_peptideHit)
+                                 std::vector < PeptideIdentification > & pep_ids,
+                                 std::vector< ProteinIdentification > & prot_ids
+                                //  Size min_n_hits_per_spectrum
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/XQuestResultXMLFile.cpp]: would remove here and handle at different part if needed

File: src/tests/class_tests/openms/source/XQuestResultXMLFile_test.cpp [hunk 11]
```diff
@@ -49,189 +46,52 @@ using namespace OpenMS;
...
-  String enzyme_name = "Trypsin";
-  Size missed_cleavages = 2;
-  String cross_link_name = "MyLinker";
-
-  String precursor_mass_tolerance_unit = "ppm";
-  String fragment_mass_tolerance_unit = "Da";
-  double precursor_mass_tolerance = 10;
-  double fragment_mass_tolerance = 0.2;
-  double fragment_mass_tolerance_xlinks = 0.3;
+  // std::cout << std::endl << "xquest xml loaded..." << std::endl;
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/XQuestResultXMLFile_test.cpp]: consider using LOG_INFO or LOG_DEBUG (or remove)

File: src/utils/OpenPepXL.cpp [hunk 12]
```diff
@@ -1171,15 +1178,17 @@ class TOPPOpenPepXL :
       {
         String precursor_mass_tolerance_unit_string = precursor_mass_tolerance_unit_ppm ? "ppm" : "Da";
         String fragment_mass_tolerance_unit_string = fragment_mass_tolerance_unit_ppm ? "ppm" : "Da";
-        XQuestResultXMLFile::writeXQuestXML(out_xquest, base_name, peptide_ids, all_top_csms, spectra,
-                                                            precursor_mass_tolerance_unit_string, fragment_mass_tolerance_unit_string, precursor_mass_tolerance, fragment_mass_tolerance, fragment_mass_tolerance_xlinks, cross_link_name,
-                                                            cross_link_mass_light, cross_link_mass_mono_link, in_fasta, in_decoy_fasta, cross_link_residue1, cross_link_residue2, cross_link_mass_iso_shift, enzyme_name, missed_cleavages);
+        // XQuestResultXMLFile::writeXQuestXML(out_xquest, base_name, peptide_ids, all_top_csms, spectra,
```
Comment (timosachsenberg) [inline src/utils/OpenPepXL.cpp]: remove?

File: src/openms/source/FORMAT/HANDLERS/XQuestResultXMLHandler.cpp [hunk 13]
```diff
@@ -63,35 +65,48 @@ namespace OpenMS
     std::map< String, UInt> XQuestResultXMLHandler::months = boost::assign::map_list_of("Jan", 1)
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/HANDLERS/XQuestResultXMLHandler.cpp]: should now work without boost
https://stackoverflow.com/questions/3250123/what-would-a-stdmap-extended-initializer-list-look-like

**Comment by timosachsenberg** [comment]
http://cdash.openms.de/viewTest.php?onlyfailed&buildid=260436

**Comment by timosachsenberg** [comment]
Looks good. I think it is read for merge if it passes ci.

**Comment by enetz** [comment]
I will try to redo the maps without boost and push all the comment and style changes together with the fixed test files

**Comment by timosachsenberg** [comment]
great! these are a lot of commits. would it hurt if I squash them into a single one before merge?

**Comment by enetz** [comment]
should be fine

File: src/topp/IDFileConverter.cpp [hunk 14]
```diff
@@ -205,10 +206,10 @@ class TOPPIDFileConverter :
     registerInputFile_("in", "<path/file>", "",
                        "Input file or directory containing the data to convert. This may be:\n"
                        "- a single file in a multi-purpose XML format (pepXML, protXML, idXML, mzid),\n"
-                       "- a single file in a search engine-specific format (Mascot: mascotXML, OMSSA: omssaXML, X! Tandem: xml, Percolator: psms),\n"
+                       "- a single file in a search engine-specific format (Mascot: mascotXML, OMSSA: omssaXML, X! Tandem: xml, Percolator: psms, xQuest: xquest.xml),\n"
```
Comment (cbielow) [inline src/topp/IDFileConverter.cpp]: just for clarity, should we prepend file extensions with a dot, i.e. write `.mascotXML, .... .xquest.xml` ?

Comment (enetz) [reply src/topp/IDFileConverter.cpp]: Yes, I agree. Will do that.

**Comment by enetz** [comment]
Looks like there are issues with pyOpenMS, had forgotten to test that.

**Comment by enetz** [comment]
only StopWatch_test failed

**Pull request merged**